### PR TITLE
added opentracing/zipkin headers support, implements #287

### LINKED ISSUE #287: OpenTracing support
This is a feature request.

## Expected Behavior
Restheart should handle incoming trace IDs (e.g. `X-B3-TraceId`) and write them to log entries where appropriate. There are multiple ways of doing it - I think keeping it simple and stupid would be enough. If a request has such an ID, write it to the logs for all log entries originating from that call.

## Current Behavior
Trace IDs are ignored.

## Context
This would make it easier to get an idea of what is happening within the system in cases where restheart is part of a larger microservice architecture (e.g. if you use a log tool like Kibana, getting all log entries of all services for one specific trace would be easy). This is mainly a feature that is useful for debugging purposes.

## Environment
n/a

## Steps to Reproduce
n/a

## Possible Implementation
You can most probably extract the `X-B3-TraceID` header when it enters your system, write that to the thread-local MDC that is used in your logger, and then simply use that in your logback configuration.

There may be more headers than just that `X-B3-TraceID`. I have seen multiple (e.g. `X-Flow-ID` and `X-Trace-ID`), but the concept has always been the same: Single header containing the trace ID, that one being written to all log entries. Maybe it makes sense to make it configurable which header should be used with a default for `X-B3-TraceID` (that one is used by the [OpenTracing initiative](http://opentracing.io/) which is being used by many people).

### PR DESCRIPTION
**Pull request by lenalebt**
This implements #287. It's a very slim implementation that will imho fit most use-cases. Incoming headers are copied to the response. If there are no headers set, we'll generate some and send them back. The `X-B3-TraceId` will be written to the logs (by default inside of the [%thread] marks).

This implementation makes it possible to correlate logged exceptions (in case this happens) with the corresponding request log, even when the thread handling the message is re-used shortly after.

It handles all headers mentioned in https://github.com/openzipkin/b3-propagation, as well as the `uber-trace-id` header that is used by  Jaeger (see https://www.jaegertracing.io/docs/client-libraries/#trace-span-identity) to get broader support.

### CODE REVIEW

**Comment by lenalebt** [comment]
Unsure about the documentation here. Users need to know that they can use `%X{name-of-the-header}` in their `logback.xml` whereever they want. I added the `x-b3-traceid` to the default `logback.xml` so normally, users might not need to think about it.

File: src/main/java/org/restheart/handlers/TracingInstrumentationHandler.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            .map(HeaderValues::peekFirst)
+            .orElse("");
+        MDC.put(X_B3_PARENTSPANID, parentSpanID);
+        String sampled = Optional.ofNullable(exchange.getRequestHeaders().get(X_B3_SAMPLED))
+            .map(HeaderValues::peekFirst)
+            .orElse("0");
+        MDC.put(X_B3_SAMPLED, sampled);
+
+        exchange.getResponseHeaders().put(HttpString.tryFromString(X_B3_TRACEID), traceID)
+            .put(HttpString.tryFromString(UBER_TRACEID), traceID)
```
Comment (lenalebt) [inline src/main/java/org/restheart/handlers/TracingInstrumentationHandler.java]: this one might need some rework to make it 100% correct (format of the trace ID for jaeger is different than for zipkin).

**Comment by mkjsix** [comment]
Hi @lenalebt 
Thank you for the PR! However, note that "This branch is out-of-date with the base branch", so you have started working from an old version. Could you please rebase your branch into master first?

**Comment by ujibang** [comment]
Hi @lenalebt,

Updating the logback.xml is not straightforward. see https://github.com/SoftInstigate/restheart/pull/203

I think that this should be controlled by few properties in the configuration file

For instance: 

```yml
requests-log-tracing-enabled: true
requests-log-tracing-header: x-b3-traceid
requests-log-tracing-id-format: jaeger
```

You can then provide some documentation directly in the configuration file in the form of configuration section comment

```yml

#### Logging

...

# Requests Tracing

# Lorem ipsum dolor sit amet, consectetur adipisci elit, sed eiusmod
# tempor incidunt ut labore et dolore magna aliqua. Ut enim ad minim
# veniam, quis nostrum exercitationem ullam corporis suscipit laboriosam
```

**Comment by lenalebt** [comment]
Okay, I think I'll change it to configurable headers (a set of them). There is opencensus.io, too - and they use different headers as well (see https://github.com/w3c/distributed-tracing/blob/master/trace_context/HTTP_HEADER_FORMAT.md).

**Comment by lenalebt** [comment]
I simplified the implementation and do not generate trace header contents anymore, since all the libs out there expect it in a different format. I refrained from creating an implementation per tracing framework.
That makes it a little less useful for cases of e.g. manual calls, but therefore easier to understand.

You need (as a user) to configure all the headers that should be taken a look at and copied to the MDC. Having an empty list will deactivate the feature.

**Comment by mkjsix** [comment]
Merged in both `master` and `3.4.x` branches. It will be available with future 3.5.0 and 3.4.2 releases.

**Comment by lenalebt** [comment]
Nice, thanks for merging :)

**Pull request merged**